server/src/main/java/org/elasticsearch/index/codec/vectors/ES814ScalarQuantizedVectorsFormat.java

```diff
@@ -0,0 +1,76 @@
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.index.codec.vectors;
+
+import org.apache.lucene.codecs.FlatVectorsFormat;
+import org.apache.lucene.codecs.FlatVectorsReader;
+import org.apache.lucene.codecs.FlatVectorsWriter;
+import org.apache.lucene.codecs.lucene99.Lucene99FlatVectorsFormat;
+import org.apache.lucene.codecs.lucene99.Lucene99ScalarQuantizedVectorsReader;
+import org.apache.lucene.index.SegmentReadState;
+import org.apache.lucene.index.SegmentWriteState;
+
+import java.io.IOException;
+
+public class ES814ScalarQuantizedVectorsFormat extends FlatVectorsFormat {
+    public static final String QUANTIZED_VECTOR_COMPONENT = "QVEC";
+
+    static final String NAME = "ES814ScalarQuantizedVectorsFormat";
+
+    static final int VERSION_START = 0;
+    static final int VERSION_CURRENT = VERSION_START;
+    static final String META_CODEC_NAME = "Lucene99ScalarQuantizedVectorsFormatMeta";
+    static final String VECTOR_DATA_CODEC_NAME = "Lucene99ScalarQuantizedVectorsFormatData";
+    static final String META_EXTENSION = "vemq";
+    static final String VECTOR_DATA_EXTENSION = "veq";
+
+    private static final FlatVectorsFormat rawVectorFormat = new Lucene99FlatVectorsFormat();
+
+    /** The minimum confidence interval */
+    private static final float MINIMUM_CONFIDENCE_INTERVAL = 0.9f;
+
+    /** The maximum confidence interval */
+    private static final float MAXIMUM_CONFIDENCE_INTERVAL = 1f;
+
+    /**
+     * Controls the confidence interval used to scalar quantize the vectors the default value is
+     * calculated as `1-1/(vector_dimensions + 1)`
+     */
+    public final Float confidenceInterval;
+
+    public ES814ScalarQuantizedVectorsFormat(Float confidenceInterval) {
+        if (confidenceInterval != null
+            && (confidenceInterval < MINIMUM_CONFIDENCE_INTERVAL || confidenceInterval > MAXIMUM_CONFIDENCE_INTERVAL)) {
+            throw new IllegalArgumentException(
+                "confidenceInterval must be between "
+                    + MINIMUM_CONFIDENCE_INTERVAL
+                    + " and "
+                    + MAXIMUM_CONFIDENCE_INTERVAL
+                    + "; confidenceInterval="
+                    + confidenceInterval
+            );
+        }
+        this.confidenceInterval = confidenceInterval;
+    }
+
+    @Override
+    public String toString() {
+        return NAME + "(name=" + NAME + ", confidenceInterval=" + confidenceInterval + ", rawVectorFormat=" + rawVectorFormat + ")";
+    }
+
+    @Override
+    public FlatVectorsWriter fieldsWriter(SegmentWriteState state) throws IOException {
+        return new ES814ScalarQuantizedVectorsWriter(state, confidenceInterval, rawVectorFormat.fieldsWriter(state));
+    }
+
+    @Override
+    public FlatVectorsReader fieldsReader(SegmentReadState state) throws IOException {
+        return new Lucene99ScalarQuantizedVectorsReader(state, rawVectorFormat.fieldsReader(state));
+    }
```

Comment:
> could you take a look at how we are using and overwriting codecs here?

The fact that we're using the reader from Lucene and our own writer looks a bit surprising, but it looks like you're suggesting forking the reader as well anyway. No concerns on my end.

Just for clarity, by forking the file format like that today, we're embarking on keeping it working for several years for bw compat reasons. (I think it's fine, but want to make sure everyone is aware.)

Reply:
> The fact that we're using the reader from Lucene and our own writer looks a bit surprising, but it looks like you're suggesting forking the reader as well anyway. No concerns on my end.

Done. https://github.com/elastic/elasticsearch/pull/106133/commits/aeb0b4a4051574431abcd2fa270d42ebb6c3a39c (if I understand correctly). It's largely pointless - since it delegate everything - and there is no materialised name relating to it, but it is otherwise benign ) 

> Just for clarity, by forking the file format like that today, we're embarking on keeping it working for several years for bw compat reasons. (I think it's fine, but want to make sure everyone is aware.)

This is a good point. Since we change the name here, then we need to support that name for ... number of releases / years. Do we really need to change the name of the format?  The format - the actual layout on disk - is not changing. The only change is to the internals of how the format is implemented.

Reply:
I'm actually less worried about this now, when we move to a future Lucene version with support for pluggable vector distance computations, then we can drop most of this implementation code. We'll need to keep the name, but the code will just all delegate to a wrapped format.